### PR TITLE
Adds unicorn project

### PR DESCRIPTION
**Pull request by catenacyber**
Waiting for https://github.com/unicorn-engine/unicorn/pull/1000 for main repo

I have a warning message when building fuzzer with memory sanitizer : 
```
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/libm.so.6
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/libpthread.so.0
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/librt.so.1
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/libdl.so.2
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/libgcc_s.so.1
WARNING: Instrumented library not found for /lib/x86_64-linux-gnu/libc.so.6
```
Is this ok ?

And it looks like a lot of undefined shifts to be fixed...

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
Does the msan build have problems running?

**Comment by eugenis** [comment]
I don't know why oss-fuzz complains, but it is completely normal for this list of libraries to be not instrumented.

**Comment by jonathanmetzman** [comment]
Thanks @eugenis 
Will merge then once we get Unicorn maintainers' approval

**Comment by Dor1s** [comment]
> I don't know why oss-fuzz complains, 

Because we provide prebuilt MSan libraries, so the warning happens when we don't have any of the libraries project needs.

@oliverchang, do we need to add those missing libraries to the MSan libs builder? I think I've seen similar warnings for some other projects too.

**Comment by eugenis** [comment]
What I'm saying is it is normal to have these libraries not instrumented. MSan has interceptors for them. We should remove the warning instead (for this exact list).

**Comment by Dor1s** [comment]
@eugenis sorry for misunderstanding and thanks for clarifying!

**Comment by catenacyber** [comment]
> Can you find a Unicorn maintainer to approve of this?

@aquynh can you approve ? (after your holidays)

And thanks for the explanation about the warnings

File: projects/unicorn/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM gcr.io/oss-fuzz-base/base-builder
+MAINTAINER unicorn.emu@gmail.com
+RUN apt-get update && apt-get install -y make python
+#TODO change to main repo
+RUN git clone --depth 1 --branch fuzz https://github.com/catenacyber/unicorn.git
```
Comment (inferno-chromium) [inline projects/unicorn/Dockerfile]: This needs to be the original unicorn repo, if there are custom patches, please add a diff file and then apply in build.sh

**Comment by inferno-chromium** [comment]
@cseagle, @lunixbochs - adding other maintainers in case they can approve.

**Comment by aquynh** [comment]
Sorry for the late reply: this looks good. Thanks everybody for doing this!

**Comment by inferno-chromium** [comment]
Thanks @aquynh 

@catenacyber - can you change to original repo in DockerFile, then we can merge this. Same for your other pull requests.

**Comment by jonathanmetzman** [comment]
@catenacyber I think you may have included some files here (eg: `projects/ecc-diff-fuzzer/build.sh`) that are not meant to be part of this PR.

**Comment by catenacyber** [comment]
Thanks @jonathanmetzman

File: projects/unicorn/build.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+cd unicorn
+./make.sh
+#we could test with make fuzz
```
Comment (jonathanmetzman) [inline projects/unicorn/build.sh]: nit: Add a space between `#` and "we"

File: projects/unicorn/build.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# build fuzz target
+cd tests/fuzz
+ls fuzz_*.c | cut -d_ -f2-4 | cut -d. -f1 | while read target
+do
+    $CC $CFLAGS -I../../include -c fuzz_$target.c -o fuzz_$target.o
+
+    $CXX $CXXFLAGS fuzz_$target.o -o $OUT/fuzz_$target ../../libunicorn.a -lFuzzingEngine
+
+    #TODO corpuses
```
Comment (jonathanmetzman) [inline projects/unicorn/build.sh]: nit: Add a space between `#` and "TODO"

**Comment by inferno-chromium** [comment]
unicorn project build is crashing on seed corpus with leaks, please fix soon or we will have to revert by EoD today.

```
Step #5: BAD BUILD: /workspace/out/address/fuzz_emu_m68k_be seems to have either startup crash or exit:
Step #5: INFO: Seed: 252797543
Step #5: INFO: Loaded 1 modules   (213646 inline 8-bit counters): 213646 [0x26f98c8, 0x272db56), 
Step #5: INFO: Loaded 1 PC tables (213646 PCs): 213646 [0x272db58,0x2a70438), 
Step #5: INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
Step #5: INFO: A corpus is not provided, starting from an empty corpus
Step #5: qemu: fatal: Illegal instruction: 0a00 @ 01000000
Step #5: AddressSanitizer:DEADLYSIGNAL
Step #5: =================================================================
Step #5: ==440==ERROR: AddressSanitizer: ABRT on unknown address 0x0000000001b8 (pc 0x7f2f50dfb428 bp 0x7ffef842c7e0 sp 0x7ffef842c568 T0)
Step #5: SCARINESS: 10 (signal)
Step #5:     #0 0x7f2f50dfb427 in gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x35427)
Step #5:     #1 0x7f2f50dfd029 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x37029)
Step #5:     #2 0x5fd02b in cpu_abort_m68k /src/unicorn/qemu/exec.c:678:5
Step #5:     #3 0xb0502b in disas_undef /src/unicorn/qemu/target-m68k/translate.c:894:5
Step #5:     #4 0xb380f1 in disas_m68k_insn /src/unicorn/qemu/target-m68k/translate.c:3058:5
Step #5:     #5 0xb26633 in gen_intermediate_code_internal_m68k /src/unicorn/qemu/target-m68k/translate.c:3149:9
Step #5:     #6 0xb26633 in gen_intermediate_code_m68k /src/unicorn/qemu/target-m68k/translate.c:3210
Step #5:     #7 0xac1e34 in cpu_m68k_gen_code /src/unicorn/qemu/translate-all.c:180:5
Step #5:     #8 0xac193f in tb_gen_code_m68k /src/unicorn/qemu/translate-all.c:1115:5
Step #5:     #9 0x1ca9948 in tb_find_slow_m68k /src/unicorn/qemu/cpu-exec.c:371:10
Step #5:     #10 0x1ca8209 in cpu_m68k_exec /src/unicorn/qemu/cpu-exec.c:211:22
Step #5:     #11 0xae6267 in tcg_exec_all_m68k /src/unicorn/qemu/cpus.c:135:17
Step #5:     #12 0xae606e in qemu_tcg_cpu_loop /src/unicorn/qemu/cpus.c:102:13
Step #5:     #13 0xae5d96 in resume_all_vcpus_m68k /src/unicorn/qemu/cpus.c:76:5
Step #5:     #14 0xae5cb0 in vm_start_m68k /src/unicorn/qemu/cpus.c:46:9
Step #5:     #15 0x532b9d in uc_emu_start /src/unicorn/qemu/../uc.c:622:9
Step #5:     #16 0x53065a in LLVMFuzzerTestOneInput /src/unicorn/tests/fuzz/fuzz_emu_m68k_be.c:48:9
Step #5:     #17 0x1f15278 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:570:15
Step #5:     #18 0x1f13117 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:479:3
Step #5:     #19 0x1f18665 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:765:5
Step #5:     #20 0x1f199b8 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:805:3
Step #5:     #21 0x1ef7d18 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:764:6
Step #5:     #22 0x1eeadcc in main /src/libfuzzer/FuzzerMain.cpp:20:10
Step #5:     #23 0x7f2f50de682f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
Step #5:     #24 0x41d328 in _start (out/address/fuzz_emu_m68k_be+0x41d328)
```

**Comment by catenacyber** [comment]
See https://github.com/unicorn-engine/unicorn/pull/1007

**Comment by inferno-chromium** [comment]
Project is disabled for now, please submit a pull request to renable once that fix is in.

**Pull request merged**